core/src/main/java/org/apache/carbondata/core/profiler/ExplainCollector.java

```diff
@@ -0,0 +1,146 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.carbondata.common.annotations.InterfaceAudience;
+import org.apache.carbondata.core.metadata.schema.table.DataMapSchema;
+
+/**
+ * An information collector used for EXPLAIN command, to print out
+ * SQL rewrite and pruning information
+ */
+@InterfaceAudience.Internal
+public class ExplainCollector {
+
+  private static final ThreadLocal<ExplainCollector> explainProfiler = new ThreadLocal<>();
+
+  private List<String> olapDataMapProviders = new ArrayList<>();
+  private List<String> olapDataMapNames = new ArrayList<>();
+
+  // mapping of table name to pruning info
+  private Map<String, TablePruningInfo> scans = new ConcurrentHashMap<>();
+
+  public void recordMatchedOlapDataMap(String dataMapProvider, String dataMapName) {
+    Objects.requireNonNull(dataMapProvider);
+    Objects.requireNonNull(dataMapName);
+    olapDataMapProviders.add(dataMapProvider);
+    olapDataMapNames.add(dataMapName);
+  }
+
+  public static boolean enabled() {
+    return explainProfiler.get() != null;
+  }
+
+  public static void setup() {
+    explainProfiler.set(new ExplainCollector());
+  }
+
+  public static ExplainCollector get() {
+    return explainProfiler.get();
+  }
+
+  public static void addPruningInfo(String tableName) {
+    if (enabled()) {
+      ExplainCollector profiler = get();
+      if (!profiler.scans.containsKey(tableName)) {
+        profiler.scans.put(tableName, new TablePruningInfo());
+      }
+    }
+  }
+
+  public static void setFilterStatement(String filterStatement) {
+    if (enabled()) {
+      TablePruningInfo scan = getCurrentTablePruningInfo();
+      scan.setFilterStatement(filterStatement);
+    }
+  }
+
+  public static void recordDefaultDataMapPruning(DataMapSchema dataMapSchema, int numBlocklets) {
+    if (enabled()) {
+      TablePruningInfo scan = getCurrentTablePruningInfo();
+      scan.setNumBlockletsAfterDefaultPruning(dataMapSchema, numBlocklets);
+    }
+  }
+
+  public static void recordCGDataMapPruning(DataMapSchema dataMapSchema, int numBlocklets) {
+    if (enabled()) {
+      TablePruningInfo scan = getCurrentTablePruningInfo();
+      scan.setNumBlockletsAfterCGPruning(dataMapSchema, numBlocklets);
+    }
+  }
+
+  public static void recordFGDataMapPruning(DataMapSchema dataMapSchema, int numBlocklets) {
+    if (enabled()) {
+      TablePruningInfo scan = getCurrentTablePruningInfo();
+      scan.setNumBlockletsAfterFGPruning(dataMapSchema, numBlocklets);
+    }
+  }
+
+  public static void setTotalBlocklets(int totalBlocklets) {
+    if (enabled()) {
+      TablePruningInfo scan = getCurrentTablePruningInfo();
+      scan.setTotalBlocklets(totalBlocklets);
+    }
+  }
+
+  /**
+   * Return the current TablePruningInfo (It is the last one in the map, since it is in
+   * single thread)
+   */
+  private static TablePruningInfo getCurrentTablePruningInfo() {
```

Comment:
ok, I will add CarbonTable parameter in all functions in this class

Reply:
Because this ExplainCollector is used when `queryExecution.toRdd.partitions` is invoked in CarbonExplainCommand. It will process relation one by one, so it is like a Stack. The last TablePruningInfo is for current table.